internal/sandbox/dirfs_unix.go

```diff
@@ -28,134 +28,149 @@ func (fsys *dirFS) Open(name string, flags int, mode fs.FileMode) (File, error)
 func (fsys *dirFS) openRoot() (File, error) {
 	dirfd, err := openat(unix.AT_FDCWD, fsys.root, O_DIRECTORY, 0)
 	if err != nil {
-		return nil, &fs.PathError{Op: "open", Path: "/", Err: err}
+		return nil, err
 	}
-	return &dirFile{fsys: fsys, fd: dirfd, name: "/"}, nil
+	return newFile(nil, dirfd), nil
+}
+
+func newFile(dir *dirFile, fd int) *dirFile {
+	f := &dirFile{dir: dir, fd: fd}
+	if dir != nil {
+		dir.ref()
+	}
+	f.ref()
+	return f
 }
 
 type dirFile struct {
-	fsys *dirFS
+	dir  *dirFile
+	refc atomic.Int32
+	once atomic.Int32
 	fd   int
-	name string
+}
+
+func (f *dirFile) ref() {
+	f.refc.Add(1)
+}
+
+func (f *dirFile) unref() {
+	if f.refc.Add(-1) == 0 {
+		closeTraceError(f.fd)
+		f.fd = -1
+		if f.dir != nil {
+			f.dir.unref()
+		}
+	}
 }
 
 func (f *dirFile) String() string {
-	return fmt.Sprintf("&sandbox.dirFile{fd:%d,name:%q}", f.fd, f.name)
+	return fmt.Sprintf("&sandbox.dirFile{fd:%d}", f.fd)
 }
 
 func (f *dirFile) Fd() uintptr {
 	return uintptr(f.fd)
 }
 
-func (f *dirFile) Name() string {
-	return f.name
-}
-
 func (f *dirFile) Close() error {
-	fd := f.fd
-	f.fd = -1
-	if fd >= 0 {
-		closeTraceError(fd)
+	if f.once.Swap(1) == 0 {
+		f.unref()
```

Comment:
What's the `once` for? Are you worried about underflow on `refc`?

Reply:
`Close` might be called multiple times, but the caller of `Open` only owns one reference to the file, so the `once` ensures that we only decrement the reference count once by closing the file.

I'm going to look at adding more validation later to prevent use of the file after Close is called as well.